### ISSUE TITLE
1.11 fails to translate IEquatable<T> implementations

### ISSUE THREAD

**Issue by jbeshir** [bridgedotnet/Bridge#1025]
### Expected

Implementations of IEquatable<T> translate successfully

### Actual

Implementations of IEquatable<T> fail to translate with an error in the form of the following message as of 1.11.1. They used to translate successfully in 1.10.4.

```
Cannot translate interface (System.IEquatable`1[[LiveApp.App+NameConflictExample]]) member '[Method LiveApp.App+NameConflictExample.Equals(x:LiveApp.App+NameConflictExample):System.Boolean]' in 'LiveApp.App+NameConflictExample' due name conflicts. Please rename methods or refactor your code
```

### Steps To Reproduce

Simply pasting this code into the "Try Live" page is sufficient to reproduce this bug.

```csharp
using Bridge;
using System;

 class App
 {
      public class NameConflictExample : IEquatable<NameConflictExample>
      {
          public override bool Equals(NameConflictExample x) { throw new NotImplementedException(); }
      }
 }
```

**Comment by DaniilVeriga** [comment]
Thank you for the report!

**Comment by vladsch** [comment]
I cannot reproduce the issue.
First, override keyword is incorrect in your test case because no suitable method to override
If remove it then test case is compiled fine

**Comment by TinkerWorX** [comment]
I can confirm that I've also run into this issue.

**Comment by vladsch** [comment]
Hi @TinkerWorX 
Can you post your test case?

**Comment by TinkerWorX** [comment]
```
   public class Operator : IEquatable<string>, IEquatable<Operator>
    {
        public Operator(string symbol, int precedence)
        {
            Symbol = symbol;
            Precedence = precedence;
        }

        public Operator(string symbol)
            : this(symbol, default(int))
        { }

        public string Symbol { get; set; }

        public int Precedence { get; set; }

        public override bool Equals(object other)
        {
            if (other is string)
                return Equals((string)other);
            if (other is Operator)
                return Equals((Operator)other);
            return false;
        }

        public bool Equals(Operator other)
        {
            return (Symbol.Equals(other.Symbol) && Precedence.Equals(other.Precedence));
        }

        public bool Equals(string other)
        {
            return Symbol.Equals(other);
        }

        public override int GetHashCode()
        {
            return Symbol.GetHashCode() + Precedence.GetHashCode() ;
        }

        public override string ToString()
        {
            return Symbol;
        }
    }

It's possible that it could be reduced more, but this reliably produces the error for me.
```

**Comment by jbeshir** [comment]
Yes, sorry about that, I spent a while experimenting in the live editor to try to get a short replication (the real code doing it is a struct with a lot of methods in ported code, which only formally implements the one IEquatable interface, but does also override the normal Equals method and operators) and apparently got confused by an error which generated a near-identical error message.

If useful I can take another try (I'm having difficulty getting the live form to reproduce it, and it dies if some parts of the original code are left in, so I'd be using a small local project). Since @TinkerWorX has put together a test case I suspect it's moot, though.

**Comment by vladsch** [comment]
@TinkerWorX 
Exception for your test is by design. Let me describe the problem
You have two Equals methods which implement IEquatable<string> and IEquatable<Operator> interfaces. In javascript, we cannot have different methods with same name. So, Bridge should emit it as equals and equals1
The problem if we assign class instance to interface variable like `IEquatable<string> i = myClass;` and try to invoke Equals method
During compilation we don't know what exactly instance in interface variable and cannot resolve Equals method name (i mean name in javascript). So, we don't know if need to emit equals1 instead equals.

Therefore Bridge raise an excepion for such cases.
We have idea how to resolve it. For example, always to use full name for methods which implement interface (as javascript name). For example,  `Equals(Operator other)` will be emitted as `Bridge.equals$1MyNamespace.Operator`

Of course, such name looks ugly in javascript but allow us to resolve such conflicts. We will review such changes to include to 1.12

Your thoughts about such change very appreciate

**Comment by jbeshir** [comment]
I think a change like that would be useful; it sounds similar to C++ name mangling, which as I understand it solves roughly the same problem of needing to turn same name, different signatures into different names deterministically so symbol resolution/linking can continue to work. Except much less ugly and not always on.

It'd be great to see something done that improves this; in my case the error breaks existing code badly enough that I can't upgrade from 1.10 for now, because there's a bunch of ported math classes being used as part of a port of an app to WebGL that're falling foul of it. Unless Bridge 2 is imminent in which case I may as well just hold out for that.

**Comment by wizzard0** [comment]
Does not work with indexers.

Repro: implement a List<T> via explicit interface implementation.
```
using System;
using System.Collections;
using System.Collections.Generic;
class C<T> : IList<T>
{
    T IList<T>.this[int index]
    {
        get
        {
            return default(T);
        }

        set
        {
            System.Console.WriteLine(value);
        }
    }

    int ICollection<T>.Count
    {
        get
        {
            throw new NotImplementedException();
        }
    }

    bool ICollection<T>.IsReadOnly
    {
        get
        {
            throw new NotImplementedException();
        }
    }

    void ICollection<T>.Add(T item)
    {
        throw new NotImplementedException();
    }

    void ICollection<T>.Clear()
    {
        throw new NotImplementedException();
    }

    bool ICollection<T>.Contains(T item)
    {
        throw new NotImplementedException();
    }

    void ICollection<T>.CopyTo(T[] array, int arrayIndex)
    {
        throw new NotImplementedException();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        throw new NotImplementedException();
    }

    IEnumerator<T> IEnumerable<T>.GetEnumerator()
    {
        throw new NotImplementedException();
    }

    int IList<T>.IndexOf(T item)
    {
        throw new NotImplementedException();
    }

    void IList<T>.Insert(int index, T item)
    {
        throw new NotImplementedException();
    }

    bool ICollection<T>.Remove(T item)
    {
        throw new NotImplementedException();
    }

    void IList<T>.RemoveAt(int index)
    {
        throw new NotImplementedException();
    }
}

```

**Comment by vladsch** [comment]
Hi @wizzard0 
Thanks for the report, cross link to the issue #1768 and pull request with the fix #1769

**Comment by LeonidVeriga** [comment]
Hi @wizzard0,

The issue #1768 mentioned [here](https://github.com/bridgedotnet/Bridge/issues/1025#issuecomment-245424630) is fixed.

**Comment by wizzard0** [comment]
That's kinda impressive reaction time, thanks!

**Comment by geoffreymcgill** [comment]
Hi @wizzard0 - Bridge 15.0 has hit NuGet and the Visual Studio Gallery. Hopefully our recent fixes for the explicit interface issue you reported are working in your project. 

If you get a chance to update your project, we would love your feedback. If you run into any issues, please let us know and we'll do our best to assist.

**Comment by geoffreymcgill** [comment]
Hi @wizzard0 - We just discovered the Bridge `ICollection<T>` class is missing `CopyTo` and `IsReadOnly`. 

We're going to add these members right away and they will be available in the **master** branch soon. We expect to release 15.1 within the next week or so.

**Comment by LeonidVeriga** [comment]
Hi @wizzard0, 

#1810 and #1811 (`IsReadOnly` and `CopyTo` in `ICollection<T>`) implemented, merged into **master** branch and will be included into `15.2`.

Your test case mentioned [here](https://github.com/bridgedotnet/Bridge/issues/1025#issuecomment-245424630) is fully supported.
